contracts/upkeeps/Cron.sol

```diff
@@ -0,0 +1,247 @@
+    2) Take the encoding, target, and handler, and create a job by sending a tx to createCronJob()
+    3) Cron job is running :)
+*/
+
+pragma solidity 0.8.6;
+
+import {CronUtility, Spec} from "../libraries/CronUtility.sol";
+import "./UpkeepBase.sol";
+import {getRevertMsg} from "../utils/utils.sol";
+
+/**
+ * @title The Cron contract
+ * @notice A keeper-compatible contract that runs various tasks on cron schedules.
+ * Users must use the encodeCronString() function to encode their cron jobs before
+ * setting them. This keeps all the string manipulation off chain and reduces gas costs.
+ */
+contract Cron is UpkeepBase {
+  //
+  using CronUtility for Spec;
+  using CronUtility for string;
+
+  event CronJobExecuted(uint256 indexed id, uint256 timestamp);
+  event CronJobCreated(uint256 indexed id, address target, bytes handler);
+  event CronJobDeleted(uint256 indexed id);
+
+  error CallFailed(uint256 id, string reason);
+  error CronJobIDNotFound(uint256 id);
+  error InvalidHandler();
+
+  uint256 private s_nextCronJobID = 1;
+  uint256[] private s_activeCronJobIDs;
+
+  mapping(uint256 => uint256) private s_lastRuns;
+  mapping(uint256 => Spec) private s_specs;
+  mapping(uint256 => address) private s_targets;
+  mapping(uint256 => bytes) private s_handlers;
+  mapping(uint256 => bytes32) private s_handlerSignatures;
+
+  /**
+   * @notice Get the id of an eligible cron job
+   * @return upkeepNeeded signals if upkeep is needed, performData is an abi encoding
+   * of the id and "next tick" of the elligible cron job
+   */
+  function checkUpkeep(bytes calldata)
+    external
+    view
+    override
+    whenNotPaused
+    returns (bool, bytes memory)
+  {
+    // DEV: start at a random spot in the list so that checks are
+    // spread evenly among cron jobs
+    uint256 idx;
+    uint256 id;
+    uint256 lastTick;
+    uint256 numCrons = s_activeCronJobIDs.length;
+    uint256 startIdx = block.number % numCrons;
+    for (idx = startIdx; idx < numCrons; idx++) {
+      id = s_activeCronJobIDs[idx];
+      lastTick = s_specs[id].lastTick();
+      if (lastTick > s_lastRuns[id]) {
+        return (true, abi.encode(id, lastTick, s_targets[id], s_handlers[id]));
+      }
+    }
+    for (idx = 0; idx < startIdx; idx++) {
+      id = s_activeCronJobIDs[idx];
+      lastTick = s_specs[id].lastTick();
+      if (lastTick > s_lastRuns[id]) {
+        return (true, abi.encode(id, lastTick, s_targets[id], s_handlers[id]));
+      }
+    }
+    return (false, bytes(""));
+  }
+
+  /**
+   * @notice Executes the cron job with id encoded in performData
+   * @param performData abi encoding of cron job ID and the cron job's next run-at datetime
+   */
+  function performUpkeep(bytes calldata performData)
+    external
+    override
+    onlyWhitelist
+    whenNotPaused
+  {
+    (uint256 id, uint256 tickTime, address target, bytes memory handler) = abi
+      .decode(performData, (uint256, uint256, address, bytes));
+    if (
```

Comment:
I'm inclined to break this if block up into three different `if`s with different error messages for readability. My personal preference would also be to take those three ifs and the one below and break them out into a private validation function.